presto-main/src/test/java/com/facebook/presto/operator/TestHashBuilderOperator.java

```diff
@@ -0,0 +1,199 @@
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.RowPagesBuilder.rowPagesBuilder;
+import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.operator.HashBuilderOperator.State.CONSUMING_INPUT;
+import static com.facebook.presto.operator.HashBuilderOperator.State.LOOKUP_SOURCE_BUILT;
+import static com.facebook.presto.operator.TestHashJoinOperator.DummySpillerFactory;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.testng.Assert.assertEquals;
+
+@Test(singleThreaded = true)
+public class TestHashBuilderOperator
+{
+    private static final SingleStreamSpillerFactory SINGLE_STREAM_SPILLER_FACTORY = new DummySpillerFactory();
+    private static final int NUM_PAGES = 1000;
+    private static final int PAGE_SIZE = 1024;
+
+    private ExecutorService executor;
+    private ScheduledExecutorService scheduledExecutor;
+
+    @BeforeMethod
+    public void setUp()
+    {
+        // Before/AfterMethod is chosen here because the executor needs to be shutdown
+        // after every single test case to terminate outstanding threads, if any.
+
+        // The line below is the same as newCachedThreadPool(daemonThreadsNamed(...)) except RejectionExecutionHandler.
+        // RejectionExecutionHandler is set to DiscardPolicy (instead of the default AbortPolicy) here.
+        // Otherwise, a large number of RejectedExecutionException will flood logging, resulting in Travis failure.
+        executor = new ThreadPoolExecutor(
+                0,
+                Integer.MAX_VALUE,
+                60L,
+                SECONDS,
+                new SynchronousQueue<Runnable>(),
+                daemonThreadsNamed("test-executor-%s"),
+                new ThreadPoolExecutor.DiscardPolicy());
+        scheduledExecutor = newScheduledThreadPool(2, daemonThreadsNamed("test-scheduledExecutor-%s"));
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void tearDown()
+            throws Exception
+    {
+        executor.shutdownNow();
+        scheduledExecutor.shutdownNow();
+        executor.awaitTermination(10, SECONDS);
+        scheduledExecutor.awaitTermination(10, SECONDS);
```

Comment:
awaitTermination returns a boolean. This needs to throw an exception if it doesn't terminate. You could invoke assertTrue.

Reply:
Actually yeah. Totally forgot about that `awaitTermination` semantics

Reply:
done.